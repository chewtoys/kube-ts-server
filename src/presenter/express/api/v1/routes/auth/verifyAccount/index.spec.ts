import dotenv from 'dotenv';
import assertOnResponseAndStatus from '../../../../../utils/tests/assertOnResponseAndStatus';
dotenv.config();
import { CONFLICT, OK } from 'http-status-codes';
import {
  API_V1,
  AUTH,
  VERIFY_ACCOUNT,
} from '../../../../../../../constants/routes';
import getUtcDate from '../../../../../../../utils/helpers/date/getUtcDate';
import usersFactory from '../../../../../utils/fakeFactories/users/factory';
import initTests from '../../../../../utils/tests/initTests';
import {
  TEST_INVALID_EMAIL,
  TEST_UUID,
  TEST_VALID_EMAIL,
} from '../../../../../utils/tests/testData';

jest.mock('uuid', () => ({
  v4: jest.fn(() => TEST_UUID),
}));

const VERIFY_ACCOUNT_URL = `${API_V1}${AUTH}${VERIFY_ACCOUNT}`;

describe('@presenter/auth/verifyAccount', () => {
  const { request, service } = initTests({ useMailServer: true });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('fails to verify account without payload', async () => {
    await assertOnResponseAndStatus({
      request,
      url: VERIFY_ACCOUNT_URL,
    });
  });

  it('fails to verify account without valid email', async () => {
    await assertOnResponseAndStatus({
      fields: {
        email: TEST_INVALID_EMAIL,
        token: TEST_UUID,
      },
      request,
      url: VERIFY_ACCOUNT_URL,
    });
  });

  it('fails to verify account without token', async () => {
    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
      },
      request,
      url: VERIFY_ACCOUNT_URL,
    });
  });

  it('fails to verify when email does not match', async () => {
    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
        token: TEST_UUID,
      },
      request,
      url: VERIFY_ACCOUNT_URL,
    });
  });

  it('fails to verify when token does not match', async () => {
    await usersFactory({
      overrides: {
        email: TEST_VALID_EMAIL,
        verifyToken: 'other-token',
      },
      service: service.users,
    });

    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
        token: TEST_UUID,
      },
      request,
      url: VERIFY_ACCOUNT_URL,
    });
  });

  it('verifies account when valid email and token is provided', async () => {
    await usersFactory({
      overrides: {
        email: TEST_VALID_EMAIL,
        id: TEST_UUID,
        verifiedAt: undefined,
        verifyToken: TEST_UUID,
      },
      service: service.users,
    });

    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
        token: TEST_UUID,
      },
      request,
      statusCode: OK,
      url: VERIFY_ACCOUNT_URL,
    });
  });

  it.only('fails to verify already verified account', async () => {
    await usersFactory({
      overrides: {
        email: TEST_VALID_EMAIL,
        id: TEST_UUID,
        verifiedAt: getUtcDate(),
        verifyToken: TEST_UUID,
      },
      service: service.users,
    });
    
    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
        token: TEST_UUID,
      },
      request,
      statusCode: CONFLICT,
      url: VERIFY_ACCOUNT_URL,
    });
  });
  // tslint:disable-next-line:max-file-line-count
});
