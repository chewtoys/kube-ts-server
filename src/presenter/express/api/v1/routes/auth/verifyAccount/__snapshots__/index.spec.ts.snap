// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@presenter/auth/verifyAccount fails to verify account without payload 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "Expected string between 0 and 255 characters long",
    },
    Object {
      "field": "token",
      "message": "Expected string 36 characters long",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/verifyAccount fails to verify account without token 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "token",
      "message": "Expected string 36 characters long",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/verifyAccount fails to verify account without valid email 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "Invalid email",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/verifyAccount fails to verify already verified account 1`] = `
Object {
  "message": "Account has been already verified",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/verifyAccount fails to verify when email does not match 1`] = `
Object {
  "message": "Invalid email and/or verify account token",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/verifyAccount fails to verify when token does not match 1`] = `
Object {
  "message": "Invalid email and/or verify account token",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/verifyAccount verifies account when valid email and token is provided 1`] = `
Object {
  "message": "Account verified successfully. You can now log in.",
}
`;
